multinsclient/client.go

```diff
@@ -0,0 +1,177 @@
+		}
+		seenNamespaces[namespace.ClientOptions.Namespace] = true
+		c.allNamespaces = append(c.allNamespaces, namespace.ClientOptions.Namespace)
+		if !namespace.Disabled {
+			c.enabledNamespaces = append(c.enabledNamespaces, namespace.ClientOptions.Namespace)
+			var err error
+			// TODO(cretz): Could alter options to have client options unrelated to
+			// namespaces and then use the new client.NewClientFromExisting introduced
+			// in https://github.com/temporalio/sdk-go/pull/881 to share connections
+			// across namespaces, but having multiple clients is mostly harmless.
+			if options.ConnectLazy {
+				c.enabledNamespaceClients[namespace.ClientOptions.Namespace], err = client.NewLazyClient(namespace.ClientOptions)
+			} else {
+				c.enabledNamespaceClients[namespace.ClientOptions.Namespace], err = client.Dial(namespace.ClientOptions)
+			}
+			if err != nil {
+				return nil, fmt.Errorf("failed creating client for namespace %q: %w", namespace.ClientOptions.Namespace, err)
+			}
+		}
+	}
+
+	// Must be at least one enabled
+	if len(c.enabledNamespaces) == 0 {
+		return nil, fmt.Errorf("no enabled namespaces")
+	}
+
+	success = true
+	return c, nil
+}
+
+// NamespaceFor selects a namespace and client for the given workflow ID. The
+// resulting namespace is always the same for the same workflow ID.
+//
+// If onlyEnabled is true and the namespace that would be selected is not
+// enabled, a hash is done only across the enabled namespace. Therefore, while
+// the same namespace is always selected even after Reset for the same workflow
+// ID when onlyEnabled is false, when onlyEnabled is true, the selected
+// namespace may be different for each Reset.
+//
+// The resulting client is only non-nil if the namespace is enabled.
+func (c *Client) NamespaceFor(workflowID string, onlyEnabled bool) (namespace string, client client.Client) {
+	// First try across all namespaces
+	namespace = c.allNamespaces[c.hasher(workflowID, len(c.allNamespaces))]
+	// If they only want enabled and this one isn't enabled, do another hash only
+	// across the enabled ones
+	if onlyEnabled && c.enabledNamespaceClients[namespace] == nil {
+		namespace = c.enabledNamespaces[c.hasher(workflowID, len(c.enabledNamespaces))]
+	}
+	client = c.enabledNamespaceClients[namespace]
+	return
+}
+
+// ClientFor is a shortcut for NamespaceFor with onlyEnabled as true.
+func (c *Client) ClientFor(workflowID string) client.Client {
+	_, client := c.NamespaceFor(workflowID, true)
+	return client
+}
+
+// Reset creates a new client from the given options. The namespaces must be the
+// same count with the same names in the same order. If closeCurrent is true,
+// this will close the current client before returning a new client on success.
+//
+// Reset is used for creating a new client if which namespaces are enabled
+// changes. This does close and reconnect all client connections.
+func (c *Client) Reset(options Options, closeCurrent bool) (*Client, error) {
```

Comment:
`Reset` sounds a bit confusing to me, and it behaves more like clone or fork. Could we instead turn it into `Update` or `UpdateConfig` that would atomically switch to the new configuration internally and close old clients always?

Reply:
`Update` sounds like it mutates the receiver. Can you think of another name?

Reply:
I think we don't need this Reset method. When the config changed, just call the New() method to get new one.

Reply:
Sure we don't need it. In fact, most of this PR is a "don't need" because it's so trivial. I think this method deserves to be here to understand the semantics of "reset".